llm-github.el

```diff
@@ -0,0 +1,56 @@
+;;; llm-github.el --- llm module for integrating with GitHub Models -*- lexical-binding: t; package-lint-main-file: "llm.el"; byte-compile-docstring-max-column: 200 -*-
+
+;; Copyright (c) 2024 Free Software Foundation, Inc.
+
+;; Author: Gabriel Santos de Souza <gabrielsantosdesouza@disroot.org>
+;; Homepage: https://github.com/ahyatt/llm
+;; SPDX-License-Identifier: GPL-3.0-or-later
+
+;; This program is free software: you can redistribute it and/or
+;; modify it under the terms of the GNU General Public License as
+;; published by the Free Software Foundation; either version 3 of the
+;; License, or (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful, but
+;; WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;; General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;; This file implements the llm functionality defined in llm.el,
+;; for the GitHub Models platform.
+
+;;; Code:
+
+(require 'llm)
+(require 'llm-openai)
+(require 'cl-lib)
+
+(cl-defstruct (llm-github (:include llm-openai-compatible (url "https://models.inference.ai.azure.com"))))
```

Comment:
> Interesting that this is an Azure URL. Can or should we base it off of the azure provider instead? Perhaps it could just be a Azure provider with a different default value for the URL.

It could be, as it also supports the Azure SDK:

<details>

![07:38 AM 18-12-2024](https://github.com/user-attachments/assets/41f3d8ef-2eab-455e-8fc6-5eb794264520)
</details>

Python Example on the page:

<details>

```python
import os
from azure.ai.inference import ChatCompletionsClient
from azure.ai.inference.models import SystemMessage, UserMessage
from azure.core.credentials import AzureKeyCredential

endpoint = "https://models.inference.ai.azure.com"
model_name = "Meta-Llama-3.1-405B-Instruct"
token = os.environ["GITHUB_TOKEN"]

client = ChatCompletionsClient(
    endpoint=endpoint,
    credential=AzureKeyCredential(token),
)

response = client.complete(
    messages=[
        SystemMessage(content="You are a helpful assistant."),
        UserMessage(content="What is the capital of France?"),
    ],
    temperature=1.0,
    top_p=1.0,
    max_tokens=1000,
    model=model_name
)

print(response.choices[0].message.content)
```
</details>

For basing it off (as in, making a new file like I did), I don't think it would be necessary, as it already works (the page on the screenshot only has the Azure example, but I was able to make a request with the current configuration).

But, if we were to expand the Azure provider to support Github Models, that would require setting the chat and embedding URL as arguments, or add a conditional to change the formatting URL.

Example with chat URL:

```diff
(cl-defmethod llm-provider-chat-url ((provider llm-azure))
-  (format "%s/openai/deployments/%s/chat/completions?api-version=2024-08-01-preview"
-          (llm-azure-url provider)
-          (llm-azure-chat-model provider)))
```
```diff
(cl-defmethod llm-provider-chat-url ((provider llm-azure))
+  (if (llm-azure-github t)
+      (format "%s/chat/completions"
+              (llm-azure-url provider))
+    (format "%s/openai/deployments/%s/chat/completions?api-version=2024-08-01-preview"
+            (llm-azure-url provider)
+            (llm-azure-chat-model provider))))
```

Reply:
You shouldn't need to do the if, though, you could have your struct include the `llm-azure` struct, and then you could keep your code as normal.

```emacs-lisp
(cl-defmethod llm-provider-chat-url ((provider llm-github))
  (format "%s/chat/completions" (llm-azure-url provider))))
```

Reply:
> You shouldn't need to do the if, though, you could have your struct include the `llm-azure` struct, and then you could keep your code as normal.

Thanks, sorry about that, I don't know much about Common Lisp.

I tried both of these:

<details>

```diff
-(require 'llm-openai)
+(require 'llm-azure)

-(cl-defstruct (llm-github (:include llm-openai-compatible (url "https://models.inference.ai.azure.com"))))
+(cl-defstruct (llm-github (:include llm-azure (url "https://models.inference.ai.azure.com"))))
```
</details>

<details>

```diff
```diff
-(require 'llm-openai)
+(require 'llm-azure)

-(cl-defstruct (llm-github (:include llm-openai-compatible (url "https://models.inference.ai.azure.com"))))
+(cl-defstruct (llm-github (:include llm-azure (url "https://models.inference.ai.azure.com"))))

 (cl-defmethod llm-provider-chat-url ((provider llm-github))
   (format "%s/chat/completions"
-          (llm-github-url provider)))
+          (llm-azure-url provider)))

 (cl-defmethod llm-provider-embedding-url ((provider llm-github) &optional _)
   (format "%s/embeddings/"
-          (llm-github-url provider)))
+          (llm-azure-url provider)))
```

But my requests timed out.